### ISSUE TITLE
Fake /proc/net contents for Docker container, enhancement

### ISSUE THREAD

**Issue by titpetric** [firehol/netdata#1238]
Hi,

as you know, docker has the unfortunate side effect of not being able to access the hosts network interfaces because of namespacing. This includes mainly the `/proc/net/dev` output, but it may be possible to get around it, as netdata it looking at this in a read-only way.

My idea: I would run a simple bash script on the host which would traverse the at least /proc/net part of the proc file system. Basically it would do something like this (pardon my pseudo code):

~~~
while [ true ]; do
   cat /proc/net/dev  > /fake/dev
   sleep 1
done
~~~

We can then mount this '/fake' location into the netdata container, where netdata doesn't have any permissions issues about reading it anymore.

Depending on the amount of files under /proc/net that are being read by netdata, it would make sense to create a volume mount separately, or if it's just the one file, to mount a volume directly to the netdata container. Currently we use `-s /host` to specify just the root location of the host proc/sys filesystems, but this is not enough. What I propose is:

- [ ] A new command line switch to specify the `/proc/net` location explicitly, separately from `/` with `-s`.
- [ ] A definitive list of files which may be subject to this kind of reading under `/proc/net`

In theory, to monitor network interface, we can then:

1. run a simple script in the background (`screen`, `nohup script &` or some other way),
2. mount a new volume with the contents of what is found under `/proc/net` on the host,
3. netdata switch for specifying an alternative `/proc/net` or `/proc/net/dev` location
4. holy network monitoring :)

This can be up and running pretty-quickly, I thought I'd just check with you instead of trying to test it out with an overlaping volume mount like `-v /proc:/host/proc -v /fakenetdev:/host/proc/net/dev` which might be very well bad practice & cause side effects.

**Comment by titpetric** [comment]
A bit of fancy grep/sed work tells me we use these files explicitly, would it be the "definite" list of files used?

~~~
     17 proc/net/dev
      6 proc/net/ip
     22 proc/net/netstat
     13 proc/net/rpc/nfs
     27 proc/net/rpc/nfsd
     56 proc/net/snmp
      8 proc/net/softnet
      1 proc/net/stat/conntrack
     10 proc/net/stat/nf
      9 proc/net/stat/synproxy
~~~

**Comment by ktsaou** [comment]
netdata already supports option `-s /path` and configuration option `[global].host access prefix = /path`.

`/path` will be used to prefix all `/proc` and `/sys` filenames.

So, `/proc/net/dev` will become `/path/proc/net/dev`.

netdata uses several files:

```sh
# egrep -o '/(proc|sys)/[a-z0-9/]+"' src/*.c | cut -d ':' -f 2 | sort -u
/proc/1/mountinfo"
/proc/diskstats"
/proc/interrupts"
/proc/loadavg"
/proc/meminfo"
/proc/net/dev"
/proc/net/netstat"
/proc/net/rpc/nfs"
/proc/net/rpc/nfsd"
/proc/net/snmp"
/proc/net/snmp6"
/proc/net/stat/conntrack"
/proc/net/stat/synproxy"
/proc/self/mountinfo"
/proc/softirqs"
/proc/stat"
/proc/vmstat"
/sys/fs/cgroup/blkio"
/sys/fs/cgroup/cpuacct"
/sys/fs/cgroup/devices"
/sys/fs/cgroup/memory"
/sys/kernel/mm/ksm"
```

**Comment by ktsaou** [comment]
Will this do?

**Comment by ktsaou** [comment]
This is a better list:

```sh
# egrep -ho '/(proc|sys)/[^[:space:]]+"' src/*.c | cut -d '"' -f 1 | sort -u
/proc/1/mountinfo
/proc/%d/cmdline
/proc/%d/fd
/proc/%d/fd/%s
/proc/%d/io
/proc/diskstats
/proc/diskstats:%s
/proc/%d/stat
/proc/%d/statm
/proc/interrupts
/proc/loadavg
/proc/meminfo
/proc/net/dev
/proc/net/dev:fireqos_monitor
/proc/net/dev:lo
/proc/net/dev:%s
/proc/net/ip_vs_stats
/proc/net/ip_vs/stats
/proc/net/netstat
/proc/net/netstat.
/proc/net/rpc/nfs
/proc/net/rpc/nfsd
/proc/net/snmp
/proc/net/snmp.
/proc/net/snmp6
/proc/net/softnet_stat
/proc/net/stat/conntrack
/proc/net/stat/nf_conntrack
/proc/net/stat/synproxy
/proc/self/mountinfo
/proc/self/oom_score_adj
/proc/softirqs
/proc/stat
/proc/sys/kernel/pid_max
/proc/sys/kernel/random/entropy_avail
/proc/vmstat
/sys/block/%s/queue/hw_sector_size
/sys/dev/block/%lu:%lu/%s
/sys/dev/block/%lu:%lu/subsystem/%s/../queue/hw_sector_size
/sys/fs/cgroup/blkio
/sys/fs/cgroup/cpuacct
/sys/fs/cgroup/devices
/sys/fs/cgroup/memory
/sys/kernel/mm/ksm
/sys/kernel/mm/ksm/pages_shared
/sys/kernel/mm/ksm/pages_sharing
/sys/kernel/mm/ksm/pages_to_scan
/sys/kernel/mm/ksm/pages_unshared
/sys/kernel/mm/ksm/pages_volatile
```

**Comment by ktsaou** [comment]
hm... this one ignore filenames in comments.
Only code:

```sh
# egrep -ho '"(%s)?/(proc|sys)/[^[:space:]]+"' src/*.c | sed 's|^"%s/|"/|' | cut -d '"' -f 2 | sort -u
/proc/1/mountinfo
/proc/%d/cmdline
/proc/%d/fd
/proc/%d/fd/%s
/proc/%d/io
/proc/diskstats
/proc/%d/stat
/proc/%d/statm
/proc/interrupts
/proc/loadavg
/proc/meminfo
/proc/net/dev
/proc/net/ip_vs_stats
/proc/net/ip_vs/stats
/proc/net/netstat
/proc/net/rpc/nfs
/proc/net/rpc/nfsd
/proc/net/snmp
/proc/net/snmp6
/proc/net/softnet_stat
/proc/net/stat/conntrack
/proc/net/stat/nf_conntrack
/proc/net/stat/synproxy
/proc/self/mountinfo
/proc/self/oom_score_adj
/proc/softirqs
/proc/stat
/proc/sys/kernel/pid_max
/proc/sys/kernel/random/entropy_avail
/proc/vmstat
/sys/block/%s/queue/hw_sector_size
/sys/dev/block/%lu:%lu/%s
/sys/dev/block/%lu:%lu/subsystem/%s/../queue/hw_sector_size
/sys/fs/cgroup/blkio
/sys/fs/cgroup/cpuacct
/sys/fs/cgroup/devices
/sys/fs/cgroup/memory
/sys/kernel/mm/ksm
/sys/kernel/mm/ksm/pages_shared
/sys/kernel/mm/ksm/pages_sharing
/sys/kernel/mm/ksm/pages_to_scan
/sys/kernel/mm/ksm/pages_unshared
/sys/kernel/mm/ksm/pages_volatile
```

**Comment by titpetric** [comment]
I'm not sure if you understood my point. As the /proc/net/ (network) is the
only part that's problematic, I would configure a seperate location for it.
For example: `-s /host` expands the above to /host/proc/* and /host/sys/*,
but i would add a new configuration option just for the '/proc/net/'
location. I'd mount this location as a volume into '/host/fakenet', pass it
to netdata as '--proc-net=/host/fakenet' (the switch would be up to you)
and you'd map the above '/proc/net/*' into '/host/fakenet/*' while leaving
the rest of /proc and /sys as it is today.

I'm not exactly clear if everything will work as expected however, because
I'm not familiar with `/proc/net/dev:lo' (what's the `:lo` part?). The
`dev` and other files in the fakenet location would be just a "copy" of
what the real proc files would output, but wouldn't exactly behave in the
same way outside of that. There would be no write controls available on the
files if some are represented by a write-read interface...

**Comment by titpetric** [comment]
Why faking the complete /proc is bad? Real time monitoring gets a penalty with such an approach, because the data is not a live representation anymore, but filled out with a specific latency.

![screenshot-192 168 1 2 8081 2016-11-12 12-00-18](https://cloud.githubusercontent.com/assets/233360/20237393/1ef9a0d0-a8d1-11e6-973b-b04355f791b0.png)

This is a screenshot with a faked net/* folder. There are a few things to note. On the left 80% of the image, there's is a jagged quality to the graph. This is because I do processing of files, and then sleep for one second. This means the files would update every 1.5 sec approximately, while netdata would read them every 1s. This results in a double-read, making the graph jagged.

At that point, I stopped the script, removed the `sleep` and restarted it. A spike is seen on the graph, after the data refreshed again, and then the jaggedness is less obvious. The reason for a smoother graph is that now the data refreshes every ~500ms, while it's beeing read every 1s. Obviously the problems would increase with faking the whole proc, but as only network namespacing is the issue, we don't have to.

When dealing with a native procfs, there is no delay of data when the files are read from. The intrinsic latency in this process makes the graph jagged, so it's best to reduce this latency as much as possible, while at the same time taking system load into account. So, my shot at best effort was:

1. read a fixed list of files (I took this from your comment)

```
/proc/net/dev
/proc/net/ip_vs_stats
/proc/net/ip_vs/stats
/proc/net/netstat
/proc/net/rpc/nfs
/proc/net/rpc/nfsd
/proc/net/snmp
/proc/net/snmp6
/proc/net/softnet_stat
/proc/net/stat/conntrack
/proc/net/stat/nf_conntrack
/proc/net/stat/synproxy
~~~

2. filter away those files which don't exist
3. store the files output on a memory media (dev/shm) to avoid disk i/o

One loop came to about 0.02 seconds, so I added the `sleep 0.08`, to have a latency of about 100ms. There's still some question if the above is a definitive list of all files we need, and if there's some way we can improve the script to use less CPU. Currently it's using about 7% of CPU which is not bad, and if I increase the sleep to 0.18/200ms latency, I end up at 3.5% on my very old atom CPU system. It might be good enough, but then again it might be better with python/other or if somebody improves my bash skills somehow.

The gist is [here](https://gist.github.com/titpetric/65edef0a4a09493f6059692deb8dbbce) if some bash/python gurus would like a stab at improving this. The script is meant to run in the background somehow (screen, nohup & fork,...). I just mounted it to the container with `-v /dev/shm/fakenet/:/host/proc/net`. I can't mount it over `-v /proc:/host/proc` because docker doesn't support this (which is my reasoning for a separate switch for just the net location).

**Comment by ktsaou** [comment]
Why don't you overwrite the settings?

```sh
FAKEPROCNET="/tmp/fake-proc-net/"; \
curl -s http://localhost:19999/netdata.conf |\
    sed "s|^[[:space:]]*#*[[:space:]]*filename to monitor[[:space:]]*=[[:space:]]*.*/proc/net/\(.*\)$|\tfilename to monitor = ${FAKEPROCNET}\1|g" >/tmp/netdata.conf.new
```

The above outputs this only computer:

```sh
# grep "filename to monitor" /tmp/netdata.conf.new
	# filename to monitor = /proc/loadavg
	# filename to monitor = /proc/interrupts
	# filename to monitor = /proc/softirqs
	filename to monitor = /tmp/fake-proc-net/softnet_stat
	# filename to monitor = /proc/sys/kernel/random/entropy_avail
	filename to monitor = /tmp/fake-proc-net/dev
	# filename to monitor = /proc/diskstats
	filename to monitor = /tmp/fake-proc-net/snmp
	filename to monitor = /tmp/fake-proc-net/snmp6
	filename to monitor = /tmp/fake-proc-net/netstat
	filename to monitor = /tmp/fake-proc-net/stat/nf_conntrack
	filename to monitor = /tmp/fake-proc-net/ip_vs_stats
	filename to monitor = /tmp/fake-proc-net/stat/synproxy
	# filename to monitor = /proc/stat
	# filename to monitor = /proc/meminfo
	# filename to monitor = /proc/vmstat
	filename to monitor = /tmp/fake-proc-net/rpc/nfsd
	filename to monitor = /tmp/fake-proc-net/rpc/nfs
```

Just copy this config to /etc/netdata/netdata.conf and restart netdata.

**Comment by ktsaou** [comment]
Keep in mind though that for this to work the next time you run it, `FAKEPROCNET=/path/` should end with `/proc/net/`. If you do this properly, you can redo this any number of times.

**Comment by titpetric** [comment]
So if I understand correctly, I can remap all the /proc/net locations in the config to new locations?

I'm struggling with netdata if somebody enables --net=host which drops isolation, I think there's not a nice way to detect it. Except maybe to count distinct interfaces as the majority of non-host docker containers have only one interface (single network). Again, there's no 100% here :/

Thanks, I'll take a look at those. If I also understand you correctly, FAKEPROCNET should contain /proc/net/ at the end? Your given example doesn't...

**Comment by ktsaou** [comment]
> So if I understand correctly, I can remap all the /proc/net locations in the config to new locations?

yes, exactly.

> Thanks, I'll take a look at those. If I also understand you correctly, FAKEPROCNET should contain /proc/net/ at the end? Your given example doesn't...

`FAKEPROCNET` needs to include `/proc/net/`, if you want the script that alters the paths (**the one I gave you above**), to re-write netdata.conf properly when run multiple times. If you never need to run the above script more than once, there is no need to have `/proc/net` in `FAKEPROCNET`. So, this is a limitation of the script I gave you above.

**Comment by titpetric** [comment]
As we're dealing with a container startup scenario (which only happens once before the container is destroyed), I don't think there's an issue with it. Of course, I can't fetch the runtime netdata config like you suggested, because the script needs to run before netdata starts. I suppose if I generate the netdata conf file only once (let's say, when building a docker image), there's no reason to update it with a running netdata, unless somebody already is doing some kind of config customisation in the netdata docker image,....

Leave the issue open for a while, I'll try to implement this in the coming week, and if it all works as expected then I'll close it myself. Thank you!

**Comment by ktsaou** [comment]
> Of course, I can't fetch the runtime netdata config like you suggested, because the script needs to run before netdata starts. I suppose if I generate the netdata conf file only once

You can have a fragment of netdata.conf which sets just the filenames you want to monitor. No need to include anything else (the default netdata.conf is commented entirely - so even an empty file will do). Then put a comment header to it, instructing the users to run:

```sh
curl http://localhost:19999/netdata.conf
```

to get a fresh copy of the file to edit. Your changes will not be lost using this procedure. The downloaded file will include them.

**Comment by titpetric** [comment]
Just to confirm, after netdata is started, I can issue a pkill -HUP netdata to reload the conf?

**Comment by ktsaou** [comment]
no, you have to restart it.

**Comment by titpetric** [comment]
I figured out SIGUSR2 also didn't do the trick :) I'm on my way to building this.

**Comment by ktsaou** [comment]
yeah. you are right.
SIGUSR2 reloads only health configuration.
Unfortunately netdata needs quite a refactoring to allow reloading its configuration (I didn't thought of this initially, so now a lot of things need to change). Sorry...

**Comment by titpetric** [comment]
No problem. Closing this issue, I added a `fakenet.sh` provider which copies possible /proc/net locations to `/dev/shm` to avoid disk I/O. It has about 250ms refresh rate to keep CPU use down, and when mounting `/dev/shm/fakenet` into `/fakenet/proc/net`, the location will be augmented with the `netdata.conf` as you suggested above. It's a bit of a "hack" but it's better than what anybody else provides in terms of network isolation. It's definitely an argument to run it in a custom network now, if people are willing to `trade-off` some accuracy on the network graphs in comparison with completely turning off network isolation with net=host.

Relevant scripts and docs have been updated in the [titpetric/netdata](https://github.com/titpetric/netdata) repository. I'm closing the issue to mark as done. Thank you!
